### PR TITLE
Prepare AP and AR to be frozen string friendly

### PR DESCRIPTION
**Pull request by kirs**
This PR makes ActionPack and ActiveRecord frozen string -friendly.
Shipping it should unblock us to move towards enforcing frozen-string-literal across all Rails components.

@matthewd 

ref https://github.com/rails/rails/pull/29540
ref https://github.com/rails/rails/pull/29506

### CODE REVIEW

**Comment by pat** [comment]
@kirs just to note that #29506 already covers off ActiveRecord, ActiveSupport, ActionMailer and ActionView (plus ActionJob didn't, when I last checked, require any changes). Glad to have ActionPack covered though :)

Guess that means just Railties is left!

**Comment by kirs** [comment]
@pat this PR fixes half of the failures that are happening on current master with  #29506 merged and `frozen_string_literal` enabled in all files. I guess there were more of them?

**Comment by pat** [comment]
Ah, must be things that aren't covered by tests? That's how I was checking things - running the tests with `RUBYOPT="--enable-frozen-string-literal"`

**Comment by kirs** [comment]
Fixed a failing test.

@pat those can be reproduced in master with:

```
cd actionpack
env RUBYOPT="--enable-frozen-string-literal" bundle exec rake test
cd ../activerecord
env RUBYOPT="--enable-frozen-string-literal" bundle exec rake test:postgresql
```

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb [hunk 1]
```diff
@@ -564,7 +564,7 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, array: nil, **) #
               end
             else
               super
-            end
+            end.dup
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: Mismerge with b86e313df53? This doesn't seem to be needed.

**Comment by kamipo** [comment]
These `dup` in AR is necessary? I tried `env RUBYOPT="--enable-frozen-string-literal" bundle exec rake test:postgresql` but only failed the `test_to_xml`. It looks like an issue for `builder` gem.

```
% RUBYOPT="--enable-frozen-string-literal" ARCONN=postgresql be ruby -w -Itest test/cases/relations_test.rb -n test_to_xml               
Using postgresql
Run options: -n test_to_xml --seed 55082

# Running:

E

Finished in 0.255228s, 3.9181 runs/s, 0.0000 assertions/s.

  1) Error:
RelationTest#test_to_xml:
RuntimeError: can't modify frozen String
    /Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.4.0/gems/builder-3.2.3/lib/builder/xmlmarkup.rb:287:in `_special'
    /Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.4.0/gems/builder-3.2.3/lib/builder/xmlmarkup.rb:254:in `instruct!'
    /Users/kamipo/src/github.com/rails/rails/activesupport/lib/active_support/core_ext/array/conversions.rb:196:in `to_xml'
    /Users/kamipo/src/github.com/rails/rails/activerecord/lib/active_record/relation/delegation.rb:39:in `to_xml'
    test/cases/relations_test.rb:83:in `block in test_to_xml'
    /Users/kamipo/src/github.com/rails/rails/activesupport/lib/active_support/testing/assertions.rb:30:in `assert_nothing_raised'
    test/cases/relations_test.rb:83:in `test_to_xml'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
RUBYOPT="--enable-frozen-string-literal" ARCONN=postgresql bundle exec ruby -  2.30s user 0.98s system 34% cpu 9.580 total
```

**Comment by kirs** [comment]
@kamipo I think that happens because `RUBYOPT="--enable-frozen-string-literal"` and `# frozen_string_literal: true` in the top of every ActiveRecord file have different effect. The first one enforces frozen strings across _all_ files and gems and the latter enforces it only in ActiveRecord.

Our goal is to enforce it in Rails codebase (https://github.com/rails/rails/pull/29540) so I think we should rely on `# frozen_string_literal: true` and not RUBYOPT to find all the spots that we want to fix. To reproduce the failures, I used the change set from #29540.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb [hunk 2]
```diff
@@ -64,7 +64,7 @@ def visit_PrimaryKeyDefinition(o)
           end
 
           def visit_ForeignKeyDefinition(o)
-            sql = <<-SQL.strip_heredoc
+            sql = <<-SQL.strip_heredoc.dup
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb]: Is this `dup` necessary? Looks like that `strip_heredoc` returns new string and we have no `strip_heredoc.dup`.
https://github.com/rails/rails/blob/v5.1.2/activesupport/lib/active_support/core_ext/string/strip.rb#L21

```
% git grep 'strip_heredoc\.dup' | wc -l
0
```

**Comment by kirs** [comment]
@kamipo you were right, that one wasn't necessary. Thanks for 👀  it! I updated the code.

**Comment by rafaelfranca** [comment]
Could you add the magic comments to the changed files?

File: actionview/lib/action_view/template.rb [hunk 3]
```diff
@@ -291,7 +292,7 @@ def #{method_name}(local_assigns, output_buffer)
         end_src
 
         # Make sure the source is in the encoding of the returned code
-        source.force_encoding(code.encoding)
+        source = source.dup.force_encoding(code.encoding)
```
Comment (matthewd) [inline actionview/lib/action_view/template.rb]: Not needed

File: actionpack/lib/action_dispatch/http/request.rb [hunk 4]
```diff
@@ -298,7 +299,7 @@ def raw_post
     # variable is already set, wrap it in a StringIO.
     def body
       if raw_post = get_header("RAW_POST_DATA")
-        raw_post.force_encoding(Encoding::BINARY)
+        raw_post = raw_post.dup.force_encoding(Encoding::BINARY)
```
Comment (matthewd) [inline actionpack/lib/action_dispatch/http/request.rb]: I'm torn on this one.. does the change belong here, or is it the caller's responsibility to provide an unfrozen value?

File: actionpack/test/abstract_unit.rb [hunk 5]
```diff
@@ -175,7 +176,7 @@ def with_autoload_path(path)
 class Rack::TestCase < ActionDispatch::IntegrationTest
   def self.testing(klass = nil)
     if klass
-      @testing = "/#{klass.name.underscore}".sub!(/_controller$/, "")
+      @testing = "/#{klass.name.underscore}".dup.sub!(/_controller$/, "")
```
Comment (deivid-rodriguez) [inline actionpack/test/abstract_unit.rb]: What about using `sub` instead of `sub!`, maybe that works better here?

**Comment by kirs** [comment]
Thanks for feedback! I addressed those.

If you think it's a good idea, I can also add these lines to `.rubocop.yml`:

```yml
Style/FrozenStringLiteralComment:
  Enabled: true
  EnforcedStyle: always
  Include:
    - 'actionpack/**/*'
    - 'activerecord/**/*'
```

However that would made the PR diff quite big to 👀 . My initial plan was to do it in a separate PR.

**Comment by matthewd** [comment]
Yeah, separate PR.

File: actionpack/lib/action_dispatch/journey/visitors.rb [hunk 6]
```diff
@@ -185,7 +186,7 @@ def terminal(node, seed)
           end
 
           def visit_GROUP(node, seed)
-            visit(node.left, seed << "(".freeze) << ")".freeze
+            visit(node.left, seed.dup << "(".freeze) << ")".freeze
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/journey/visitors.rb]: I think we can remove all those freeze here

Comment (kirs) [inline actionpack/lib/action_dispatch/journey/visitors.rb]: Good idea! I didn't want to do that at first, but now when we have `# frozen_string_literal: true` in the updated files it makes sense to do it.

I updated the PR.

**Comment by kirs** [comment]
@matthewd please let me know if there's anything else I can improve in this PR.

**Pull request merged**